File: gisrenderer/app/src/main/java/com/nextgis/gismaplib/messages/MapboxMsg.java
package com.nextgis.gismaplib.messages;

import android.graphics.Bitmap;
import com.nextgis.gismaplib.capnproto.GmConst;
import com.nextgis.gismaplib.capnproto.GmMsg;
import org.capnproto.AnyPointer;

import java.io.IOException;


public class MapboxMsg
    extends GmCoreMsg
{
  protected static final String TAG = MapboxMsg.class.getName();

  protected Bitmap mDstBitmap;
  protected Boolean mIsImageStored        = null;
  protected boolean mIsBitmapPixelsLocked = false;


  public MapboxMsg(Bitmap dstBitmap)
  {
    super();
    mMsgType = GmConst.MSG_TYPE_MAPBOX_IMAGE_DATA;
    mDstBitmap = dstBitmap;
  }


  @Override
  protected void setDataQ(AnyPointer.Builder dataPtrQ)
  {
    // set the query data
    GmMsg.MapboxQ.Builder dataQ = dataPtrQ.initAs(GmMsg.MapboxQ.factory);

    long imagePointer = GmCoreMsg.lockBitmapPixels(mDstBitmap);
    dataQ.setImagePointer(imagePointer);

    if (imagePointer != 0) {
      mIsBitmapPixelsLocked = true;
      dataQ.setImageWidth(mDstBitmap.getWidth());
      dataQ.setImageHeight(mDstBitmap.getHeight());
    }
  }


  @Override
  protected AnyPointer.Reader msgWorker()
      throws IOException
  {
    // get the reply data
    AnyPointer.Reader dataPtrR = super.msgWorker();
    GmMsg.MapboxR.Reader dataR = dataPtrR.getAs(GmMsg.MapboxR.factory);

    mIsImageStored = dataR.getIsImageStored();
    if (mIsImageStored) {
      GmCoreMsg.unlockBitmapPixels(mDstBitmap);
      mIsBitmapPixelsLocked = false;
    }

    return (dataPtrR);
  }


  public Boolean isImageStored()
      throws IOException
  {
    if (null == mIsImageStored) {
      msgWorker();
    }
    return (mIsImageStored);
  }
}

File: gisrenderer/gismaplib/src/gismaplib/messages/MapboxMsg.cpp
#include "gismaplib/messages/MapboxMsg.h"

#include "gismaplib/GmLog.h"
//#include "gismaplib/platforms/android/GlOfflineRenderer.h"
#include "gismaplib/platforms/android/mapbox/MapboxRenderer.h"

namespace gmcore
{
  void MapboxMsg::dataWorker(
      boost::shared_ptr <capnp::AnyPointer::Reader> dataPtrQ,
      boost::shared_ptr <GmMsg::Message::Builder> msgPtrR)
  {
    // get the query data
    GmMsg::MapboxQ::Reader dataQ = dataPtrQ->getAs <GmMsg::MapboxQ>();

    uintptr_t imagePointer = static_cast<uintptr_t> (dataQ.getImagePointer());

    capnp::AnyPointer::Builder dataPtrR = msgPtrR->initDataPointer();
    GmMsg::MapboxR::Builder dataR = dataPtrR.initAs <GmMsg::MapboxR>();
    if (!imagePointer) {
      dataR.setIsImageStored(false);
      return;
    }

    void* imagePtr = reinterpret_cast<void*> (imagePointer);
    uint16_t imageWidth = (uint16_t) (dataQ.getImageWidth());
    uint16_t imageHeight = (uint16_t) (dataQ.getImageHeight());

    // get and put image data
//    GlOfflineRenderer renderer{imagePtr, imageWidth, imageHeight};
//    renderer.render();
    MapboxRenderer renderer{imagePtr, imageWidth, imageHeight};
    renderer.render();

    // set the reply data
    dataR.setIsImageStored(true);
  }
}  // namespace gmcore
